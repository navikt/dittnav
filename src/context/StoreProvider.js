import React, { createContext, useReducer } from 'react';
import { node } from 'prop-types';
import storeReducer, { initialStoreState } from '../reducers/storeReducer';

export const StoreContext = createContext(null);

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(storeReducer, initialStoreState());

  const addBeskjeder = (nyeBeskjeder) => dispatch({
    type: 'ADD_BESKJEDER',
    payload: nyeBeskjeder,
  });

  const addInaktiveBeskjeder = (nyeBeskjeder) => dispatch({
    type: 'ADD_INAKTIVE_BESKJEDER',
    payload: nyeBeskjeder,
  });

  const removeBeskjed = (beskjed) => dispatch({
    type: 'REMOVE_BESKJED',
    payload: beskjed,
  });

  const addInaktivBeskjed = (beskjed) => dispatch({
    type: 'ADD_INAKTIV_BESKJED',
    payload: beskjed,
  });

  const toggleInnloggingsModal = () => dispatch({
    type: 'TOGGLE_INNLOGGINGSMODAL',
  });

  const setError = () => dispatch({
    type: 'SET_ERROR',
  });

  return (
    <StoreContext.Provider value={{
      state,
      addBeskjeder,
      addInaktivBeskjed,
      addInaktiveBeskjeder,
      removeBeskjed,
      toggleInnloggingsModal,
      setError,
    }}
    >
      {children}
    </StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: node.isRequired,
};

export default StoreProvider;
