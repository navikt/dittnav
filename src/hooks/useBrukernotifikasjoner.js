import { useQuery } from 'react-query';
import { fetchBeskjeder, fetchOppgaver, fetchInnbokser } from '../Api';
import useStore from './useStore';

const useBeskjeder = () => {
  const { state, addBeskjeder, setError } = useStore();

  useQuery('beskjeder', fetchBeskjeder, {
    onSuccess: addBeskjeder,
    onError: setError,
  });

  return state.beskjeder;
};

const useOppgaver = () => (
  useQuery('oppgaver', fetchOppgaver, { onError: useStore().setError })
);

const useInnbokser = () => (
  useQuery('innbokser', fetchInnbokser, { onError: useStore().setError })
);

const useBrukernotifikasjoner = () => (
  [useBeskjeder(), useOppgaver(), useInnbokser()]
);

export default useBrukernotifikasjoner;
