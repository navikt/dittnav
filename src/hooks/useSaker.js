import { useQuery } from 'react-query';
import { fetchSakstema, fetchSaker } from '../Api';
import useStore from './useStore';

export const useSakstema = () => (
  [useQuery('sakstema', fetchSakstema, { onError: useStore().setError })]
);

export const usePaabegynteSoknader = () => {
  const store = useStore();

  const checkForFeilendeBaksystem = (data) => {
    if (data.content.feilendeBaksystem.length > 0) {
      store.setError();
    }
  };

  return [useQuery('paabegynteSoknader', fetchSaker, { onSuccess: checkForFeilendeBaksystem, onError: store.setError })];
};

const useSaker = () => (
  [useSakstema(), usePaabegynteSoknader()]
);

export default useSaker;
