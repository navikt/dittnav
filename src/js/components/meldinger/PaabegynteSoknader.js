import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage as F, injectIntl, intlShape } from 'react-intl';
import i18n from '../../../translations/i18n';
import { IkonInformasjon, LenkepanelMedIkon } from '../paneler/LenkepanelMedIkon';
import PanelOverskrift from '../paneler/PanelOverskrift';

const getOverskrift = (paabegynteSoknader, soknadstekst, intl) => (
  <PanelOverskrift
    overskrift={<F id={soknadstekst} values={{ count: i18n[intl.locale].numberToWord(paabegynteSoknader.antallPaabegynte) }} />}
    type="Element"
  />
);

const PaabegynteSoknader = ({ paabegynteSoknader, intl }) => {
  if (!paabegynteSoknader || paabegynteSoknader.antallPaabegynte === 0) {
    return null;
  }
  const soknadstekst = paabegynteSoknader.antallPaabegynte === 1 ? 'saksoversikt.soknad.en' : 'saksoversikt.soknad.flere';

  return (
    <LenkepanelMedIkon
      className="infoMelding"
      data-ga="Dittnav/Varsel/Paabegynt soknad"
      alt="Melding om SÃ¸knader"
      overskrift={getOverskrift(paabegynteSoknader, soknadstekst, intl)}
      ingress={<F id="saksoversikt.lenke" />}
      href={paabegynteSoknader.url}
    >
      <IkonInformasjon />
    </LenkepanelMedIkon>
  );
};

export const PaabegynteSoknaderType = PropTypes.shape({
  url: PropTypes.string.isRequired,
  antallPaabegynte: PropTypes.number.isRequired,
});

PaabegynteSoknader.propTypes = {
  intl: intlShape.isRequired, // eslint-disable-line react/no-typos
  paabegynteSoknader: PaabegynteSoknaderType,
};

PaabegynteSoknader.defaultProps = {
  paabegynteSoknader: null,
};

export default injectIntl(PaabegynteSoknader);
