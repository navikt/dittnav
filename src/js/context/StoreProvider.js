import React, { createContext, useReducer } from 'react';
import { node } from 'prop-types';
import storeReducer, { initialStoreState } from '../reducers/storeReducer';

export const StoreContext = createContext(null);

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(storeReducer, initialStoreState());

  const addBeskjeder = (nyeBeskjeder) => dispatch({
    type: 'ADD_BESKJEDER',
    payload: nyeBeskjeder,
  });

  const addInaktiveBeskjeder = (nyeBeskjeder) => dispatch({
    type: 'ADD_INAKTIVE_BESKJEDER',
    payload: nyeBeskjeder,
  });

  const removeBeskjed = (beskjed) => dispatch({
    type: 'REMOVE_BESKJED',
    payload: beskjed,
  });

  const addInaktivBeskjed = (beskjed) => dispatch({
    type: 'ADD_INAKTIV_BESKJED',
    payload: beskjed,
  });

  const toggleInnloggingsModal = () => dispatch({
    type: 'TOGGLE_INNLOGGINGSMODAL',
  });

  const addOppgaver = (oppgaver) => dispatch({
    type: 'ADD_OPPGAVER',
    payload: oppgaver,
  });

  const addInnbokser = (innbokser) => dispatch({
    type: 'ADD_INNBOKSER',
    payload: innbokser,
  });

  const addInaktiveOppgaver = (inaktiveOppgaver) => dispatch({
    type: 'ADD_INAKTIVE_OPPGAVER',
    payload: inaktiveOppgaver,
  });

  const addInaktiveInnbokser = (inaktiveInnbokser) => dispatch({
    type: 'ADD_INAKTIVE_INNBOKSER',
    payload: inaktiveInnbokser,
  });

  const addInnloggingsstatus = (innloggingsstatus) => dispatch({
    type: 'ADD_INNLOGGINGSSTATUS',
    payload: innloggingsstatus,
  });

  const addOppfolging = (oppfolging) => dispatch({
    type: 'ADD_OPPFOLGING',
    payload: oppfolging,
  });

  const addMeldekort = (meldekort) => dispatch({
    type: 'ADD_MELDEKORT',
    payload: meldekort,
  });

  const addNavn = (navn) => dispatch({
    type: 'ADD_NAVN',
    payload: navn,
  });

  const addIdent = (Ident) => dispatch({
    type: 'ADD_IDENT',
    payload: Ident,
  });

  const addPaabegynteSoknader = (paabegynteSoknader) => dispatch({
    type: 'ADD_PAABEGYNTESOKNADER',
    payload: paabegynteSoknader,
  });

  const addMeldinger = (meldinger) => dispatch({
    type: 'ADD_MELDINGER',
    payload: meldinger,
  });

  const addSakstema = (sakstema) => dispatch({
    type: 'ADD_SAKSTEMA',
    payload: sakstema,
  });

  const setOppfolgingFailed = () => dispatch({
    type: 'OPPFOLGING_FAILED',
  });

  const setError = () => dispatch({
    type: 'ERROR',
  });

  return (
    <StoreContext.Provider value={{
      state,
      addBeskjeder,
      addOppgaver,
      addInnbokser,
      addInaktivBeskjed,
      addInaktiveBeskjeder,
      addInnloggingsstatus,
      addInaktiveOppgaver,
      addInaktiveInnbokser,
      addOppfolging,
      addMeldekort,
      addNavn,
      addIdent,
      addPaabegynteSoknader,
      addMeldinger,
      addSakstema,
      removeBeskjed,
      toggleInnloggingsModal,
      setOppfolgingFailed,
      setError,
    }}
    >
      {children}
    </StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: node.isRequired,
};

export default StoreProvider;
