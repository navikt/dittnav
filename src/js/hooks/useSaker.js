import { useQuery } from 'react-query';
import Api from '../Api';
import useStore from './useStore';

export const useSakstema = () => (
  [useQuery('sakstema', Api.fetchSakstema, { onError: useStore().setError })]
);

export const usePaabegynteSoknader = () => (
  [useQuery('paabegynteSoknader', Api.fetchSaker, { onError: useStore().setError })]
);

const useSaker = () => (
  [useSakstema(), usePaabegynteSoknader()]
);

export default useSaker;
