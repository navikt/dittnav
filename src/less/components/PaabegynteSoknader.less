.paabegynte-soknader {
  margin-bottom: 1em;
  padding: 0.75em 0.75em;
  display: flex;
  flex-direction: row;
  background-color: #f5fcfd !important;
  border: 1px solid @navLysBlaLighten40 !important;

  .lenkepanel__ikon {
    align-self: center;

    svg {
      vertical-align: middle;
    }
  }
}

@media screen and (max-width: @mobile-breakpoint) {
  .paabegynte-soknader {
    padding: 1.025em 1.025em;
    padding-top: 1.25em !important;
    padding-bottom: 1.25em !important;


    .lenkepanel__ikon {
      display: none !important;
    }

    &__tekst {
      padding-left: 0;
    }
  }
}
