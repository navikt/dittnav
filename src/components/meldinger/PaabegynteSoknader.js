import React from 'react';
import { FormattedMessage as F, useIntl } from 'react-intl';
import i18n from '../../language/i18n';
import LenkepanelMedIkon from '../common/LenkepanelMedIkon';
import PanelOverskrift from '../common/PanelOverskrift';
import IkonBeskjed from '../../assets/IkonBeskjed';
import { GoogleAnalyticsAction, GoogleAnalyticsCategory } from '../../utils/googleAnalytics';
import { usePaabegynteSoknader } from '../../hooks/useSaker';

const hasContent = (paabegynteSoknader) => (
  paabegynteSoknader && paabegynteSoknader.content
);

const hasNoPaabegynteSoknader = (paabegynteSoknader) => (
  paabegynteSoknader && paabegynteSoknader.content && paabegynteSoknader.content.antallPaabegynte === 0
);

const createOverskrift = (paabegynteSoknader, soknadstekst, intl) => (
  <PanelOverskrift overskrift={<F id={soknadstekst} values={{ count: i18n[intl.locale].numberToWord(paabegynteSoknader.antallPaabegynte) }} />} type="Element" />
);

const PaabegynteSoknader = () => {
  const [{ data: paabegynteSoknader, isSuccess }] = usePaabegynteSoknader();
  const intl = useIntl();

  if (!isSuccess || !hasContent(paabegynteSoknader) || hasNoPaabegynteSoknader(paabegynteSoknader)) {
    return null;
  }

  const soknadstekst = paabegynteSoknader && paabegynteSoknader.content.antallPaabegynte === 1 ? 'saksoversikt.soknad.en' : 'saksoversikt.soknad.flere';

  return (
    <LenkepanelMedIkon
      className="infomelding paabegynte-soknader"
      alt="Melding om SÃ¸knader"
      overskrift={createOverskrift(paabegynteSoknader.content, soknadstekst, intl)}
      ingress={<F id="saksoversikt.lenke" />}
      href={paabegynteSoknader.content.url}
      gaCategory={GoogleAnalyticsCategory.Forside}
      gaAction={GoogleAnalyticsAction.PaabegynteSoknader}
    >
      <IkonBeskjed />
    </LenkepanelMedIkon>
  );
};

export default PaabegynteSoknader;
