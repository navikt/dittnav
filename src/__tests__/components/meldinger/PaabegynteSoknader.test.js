import * as React from 'react';
import sinon from 'sinon';
import PaabegynteSoknader from 'js/components/meldinger/PaabegynteSoknader';
import wrapIntl from 'js/IntlTestHelper';
import * as useSaker from '../../../js/hooks/useSaker';
import StoreProvider from '../../../js/context/StoreProvider';
const ReactTestRenderer = require('react-test-renderer');

jest.mock('react-ga');
let sandbox = null;

beforeEach(() => {
  sandbox = sinon.createSandbox();
});

afterEach(() => {
  useSaker.usePaabegynteSoknader.restore();
});

test('PaabegynteSoknader empty array', () => {
  sandbox.stub(useSaker, 'usePaabegynteSoknader')
    .returns([{
      data: { content: null },
      isLoading: false,
      isSuccess: true,
    }]);

  const component = ReactTestRenderer.create(wrapIntl(
    <StoreProvider>
      <PaabegynteSoknader />
    </StoreProvider>
  ));

  expect(component.toJSON()).toMatchSnapshot();
});

test('PaabegynteSoknader one soknad', () => {
  const paabegynteSoknader = {
    "url": "http://nav.no",
    "antallPaabegynte": 1
  };

  sandbox.stub(useSaker, 'usePaabegynteSoknader')
    .returns([{
      data: { content: paabegynteSoknader },
      isLoading: false,
      isSuccess: true,
    }]);

  const component = ReactTestRenderer.create(wrapIntl(
    <StoreProvider>
      <PaabegynteSoknader />
    </StoreProvider>
  ));

  expect(component.toJSON()).toMatchSnapshot();
});

test('PaabegynteSoknader several soknader', () => {
  const paabegynteSoknader = {
    "url": "https://tjenester-t6.nav.no/",
    "antallPaabegynte": 3
  };

  sandbox.stub(useSaker, 'usePaabegynteSoknader')
    .returns([{
      data: { content: paabegynteSoknader },
      isLoading: false,
      isSuccess: true,
    }]);

  const component = ReactTestRenderer.create(wrapIntl(
    <StoreProvider>
      <PaabegynteSoknader />
    </StoreProvider>
    ));

  expect(component.toJSON()).toMatchSnapshot();
});
