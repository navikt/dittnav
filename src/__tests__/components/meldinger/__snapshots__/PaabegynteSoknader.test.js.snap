// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaabegynteSoknader empty array 1`] = `null`;

exports[`PaabegynteSoknader green test 1`] = `null`;

exports[`PaabegynteSoknader one soknad 1`] = `
<a
  className="message clickable"
  data-ga="Dittnav/Varsel/Paabegynt soknad"
  href="http://nav.no"
>
  <span
    aria-label="dokument-ikon"
    className="icon document-icon"
  />
  <div
    className="texts"
  >
    <p>
      Du har en søknad som du ikke har sendt
    </p>
    <p
      id="paabegynte-tekst"
    >
      Fortsett innsending
    </p>
  </div>
</a>
`;

exports[`PaabegynteSoknader several soknad 1`] = `
<a
  className="message clickable"
  data-ga="Dittnav/Varsel/Paabegynt soknad"
  href="/saksoversikt"
>
  <span
    aria-label="dokument-ikon"
    className="icon document-icon"
  />
  <div
    className="texts"
  >
    <p>
      Du har to søknader som du ikke har sendt
    </p>
    <p
      id="paabegynte-tekst"
    >
      Fortsett innsending
    </p>
  </div>
</a>
`;
